### PR TITLE
header standardization

### PR DESCRIPTION
**Pull request by fil512**
New interfaces and utilities have been added to convert the X-Request-Partition-IDs header into a
RequestPartitionId. Also a new IDefaultPartitionSettings interface has been added to provide default partition
details to assist in converting X-Request-Partition-IDs=DEFAULT into a default partition id.

### CODE REVIEW

**Comment by robogary** [comment]
Formatting check succeeded!

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/Constants.java [hunk 1]
```diff
@@ -33,25 +33,97 @@
...
+	public static final String HEADER_X_SECURITY_CONTEXT = "X-Security-Context";
+	public static final String HEADER_CASCADE = "X-Cascade";
+	public static final String HEADER_X_PROGRESS = "X-Progress";
+	public static final String HEADER_X_PROGRESS_LC = HEADER_X_PROGRESS.toLowerCase();
+	public static final String HEADER_X_FHIR_STARTER = "X-FHIR-Starter";
+	public static final String HEADER_X_REQUEST_PARTITION_IDS = "X-Request-Partition-IDs";
+
+	// ==============================================================
+	// Other Constants
+	// ==============================================================
```
Comment (TipzCM) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/api/Constants.java:118]: i love these little headers

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	 *
+	 * @param theSourceName               The name of the source of the value, used for error messages
+	 * @param thePartitionHeaderValue     The value of the X-Request-Partition-IDs header, may be null
+	 * @param theIncludeOnlyTheFirst      If true, only the first partition ID in the header will be included in the result
+	 * @param theDefaultPartitionSettings Settings that provide the default partition ID
+	 * @return A {@link RequestPartitionId} object representing the partition(s) specified in the header, or null if the header is null
+	 * @throws InvalidRequestException If the header value is invalid
+	 */
+	@Nullable
+	private static RequestPartitionId fromHeader(
```
Comment (TipzCM) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:121]: I can always tell when you use the refactor tools because it puts static and annotations on the method.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	 *
+	 * @param <T>                         The type of the payload, which must have methods for getting and setting a partition ID
+	 * @param theMessage                  The message containing the payload and headers
+	 * @param theDefaultPartitionSettings Settings that provide the default partition ID
+	 */
+	public static <T> void setRequestPartitionIdFromHeaderIfNotAlreadySet(
+			@Nonnull IMessage<T> theMessage, @Nonnull IDefaultPartitionSettings theDefaultPartitionSettings) {
+		if (theMessage.getPayload() instanceof BaseResourceMessage baseResourceMessage) {
+			if (baseResourceMessage.getPartitionId() != null) {
+				// TODO KHS suggestion from MB: if partitions are also set in the header, log a warning if they don't
```
Comment (TipzCM) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:210]: Do we want to do this?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:121]: We've had this conversation before. I make the method static when it's static :-)

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:210]: Not yet. It feels expensive for a rare edge case. The precedence is documented.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	/**
+	 * Test whether <code>theRequestPartitionId</code> is only targeting the default partition where the ID of the default
+	 * partition is provided by {@link #getDefaultPartitionId()}.
+	 *
+	 * @param theRequestPartitionId to perform the evaluation upon.
+	 * @return true if the <code>theRequestPartitionId</code> is for the default partition only.
+	 */
+	default boolean isDefaultPartition(@Nonnull RequestPartitionId theRequestPartitionId) {
+		return theRequestPartitionId.isDefaultPartition(getDefaultPartitionId());
+	}
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java:27]: issue: I do not understand this method. We receive a RequestPartitionId, and then, given our _own_ knowledge of the default partition ID in this class, we ask the requestpartition object to tell if it is our default, by passing it our default? Why not just call `theRequestPartitionId.isDefaultPartition(myDefaultPartitionSettings.getDefaultPartition())` and remove this method entirely? 

IMO this would be better refactored into a method called `theRequestPartitionId.isForPartition(getDefaultPartitionId()`. This asker asking another service to ask itself, is a pattern that constitutes a code smell I believe.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	 * @param thePartitionHeaderValue     The value of the X-Request-Partition-IDs header, may be null
+	 * @param theDefaultPartitionSettings Settings that provide the default partition ID
+	 * @return A {@link RequestPartitionId} object representing the partition(s) specified in the header, or null if the header is null
+	 * @throws InvalidRequestException If the header value is invalid
+	 */
+	@Nullable
+	public static RequestPartitionId fromHeader(
+			@Nonnull String theSourceName,
+			@Nullable String thePartitionHeaderValue,
+			@Nullable IDefaultPartitionSettings theDefaultPartitionSettings) {
```
Comment (tadgh) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:45]: question: should IDefaultPartitionSettings be nullable? (note, i ask this once, but it appears in many places)

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/HapiMessageHeaders.java [hunk 6]
```diff
@@ -106,4 +106,15 @@ public MessageHeaders toMessageHeaders() {
...
+
+	/**
+	 * Adds a custom header to the message headers.
+	 * This method allows adding arbitrary key-value pairs to the custom headers map.
+	 *
+	 * @param theKey the header key/name
+	 * @param theValue the header value
+	 */
+	public void addCustomHeader(String theKey, Object theValue) {
+		getCustomHeaders().put(theKey, theValue);
```
Comment (tadgh) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/HapiMessageHeaders.java:118]: issue: the underlying header is not a multimap, but probably should be.most implementations of headers use a multimap of some sort, that is Map<String, List<String>>

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/broker/jms/SpringMessagingMessageHandlerAdapter.java [hunk 7]
```diff
@@ -52,7 +52,11 @@ public void handleMessage(Message<?> theMessage) throws MessagingException {
 		if (theMessage instanceof GenericMessage genericMessage) {
 			// When receiving a message from an external queue, it will likely arrive as a GenericMessage
 			Object payload = genericMessage.getPayload();
-			message = new RawStringMessage(genericMessage.getPayload().toString(), genericMessage.getHeaders());
+			if (payload instanceof byte[] bytes) {
+				message = new RawStringMessage(new String(bytes), genericMessage.getHeaders());
+			} else {
+				message = new RawStringMessage(payload.toString(), genericMessage.getHeaders());
```
Comment (tadgh) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/broker/jms/SpringMessagingMessageHandlerAdapter.java:58]: question: why this change?

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/RequestHeaderPartitionInterceptor.java [hunk 8]
```diff
@@ -22,104 +22,74 @@
...
+			RequestDetails theRequestDetails,
+			BiFunction<String, IDefaultPartitionSettings, RequestPartitionId> aHeaderParser) {
+		String partitionHeader = theRequestDetails.getHeader(Constants.HEADER_X_REQUEST_PARTITION_IDS);
 
-			// return early if we only need the first partition ID
-			if (theIncludeOnlyTheFirst) {
-				return RequestPartitionId.fromPartitionIds(partitionIds);
+		if (isBlank(partitionHeader)) {
+			if (theRequestDetails instanceof SystemRequestDetails) {
+				return myDefaultPartitionSettings.getDefaultRequestPartitionId();
```
Comment (tadgh) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/RequestHeaderPartitionInterceptor.java]: Question: Are we sure that all usages of SystemRequestDetails are now using this method? If memory serves it could set a request partition ID manually (though my memory of this may be wrong).

**Review by tadgh — APPROVED** [review]
Approved. Wuuld be nice to fix that wonky API, but its probably a fairly large lift, so can be delayed to the future.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java (hunk 5 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/RequestPartitionHeaderUtil.java:45]: It was `@Nonnull` until Leif's review when he pointed out there will be contexts where this is legit null so I updated the code to handle that case.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/HapiMessageHeaders.java (hunk 6 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/HapiMessageHeaders.java:118]: Good point. However, this code has been like this for years and a lot of stuff runs on it. I'd be nervous about changing the behaviour. E.g. maybe something is depending on headers.put("id", 1) followed by headers.put("id", 2) to overwrite the value.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/broker/jms/SpringMessagingMessageHandlerAdapter.java (hunk 7 above)
Comment (fil512) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/broker/jms/SpringMessagingMessageHandlerAdapter.java:58]: Good question. This was a case that Adi faced.

Comment (fil512) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/broker/jms/SpringMessagingMessageHandlerAdapter.java:58]: Something used to work and stopped working when I did the big message api refactor, and this fixed that regression. I think it might have been an ActiveMq consumer for TransactionLog events that were sent by our JMS Transaction log publisher. Something like that. There's a test in cdr that this fixes.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/RequestHeaderPartitionInterceptor.java (hunk 8 above)
Comment (fil512) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/RequestHeaderPartitionInterceptor.java]: Very good point. Fixing now.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java (hunk 4 above)
Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java:27]: I have a service that knows all about the default partition. I have a partition in my hand. I want to know if it's the default partition. I would expect to be able to call a method defaultPartitionService.isDefaultPartition(myPartition). It's just a convenience method. It's not clear to me what you don't like about it. Note there is also a hasDefaultPartition which does something subtly different

Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java:27]: I take your point on the theRequestPartitionId.isDefaultPartition method being misnamed. I will fix that now

Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java:27]: fixed the name of the request partition id. Hope you're good with this comprimise

Comment (tadgh) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/model/IDefaultPartitionSettings.java:27]: sure good enough :+1: we can chat later, maybe im just not understanding it.
